internal/config/server/crypto.go

```diff
@@ -0,0 +1,48 @@
+// Copyright 2024 Stacklok, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+// CryptoConfig is the configuration for the crypto engine
+type CryptoConfig struct {
+	KeyStore KeyStoreConfig `mapstructure:"key_store"`
+	Default  DefaultCrypto  `mapstructure:"default"`
+	Fallback FallbackCrypto `mapstructure:"fallback"`
+}
+
+// KeyStoreConfig specifies the type of keystore to use and its configuration
+type KeyStoreConfig struct {
+	Type string `mapstructure:"type"`
+	// is currently expected to match the structure of LocalFileKeyStoreConfig
+	Config map[string]any `mapstructure:"config"`
+}
+
+// DefaultCrypto defines the default cyrpto to be used for new data
+type DefaultCrypto struct {
+	KeyID     string `mapstructure:"key_id"`
+	Algorithm string `mapstructure:"algorithm"`
+}
+
+// FallbackCrypto defines the optional list of keys and algorithms to fall
+// back to.
+// When rotating keys or algorithms, add the old ones here.
+type FallbackCrypto struct {
+	KeyIDs     []string `mapstructure:"key_ids"`
+	Algorithms []string `mapstructure:"algorithms"`
```

Comment:
Should this be a `[]struct{KeyID string, Algo string}`?  I'm wondering what happens if the two arrays are not the same length.

Reply:
It is expected that they will not always be the same length. For example, if we rotate a key, but do not change the algorithm, then we will add a fallback key but not a corresponding algorithm.

The selection of key/algorithm during decryption is based on the data we store in the database (see the `EncryptedData` struct in `models.go`) and the selection of the algorithm during encryption is based on the values in the `DefaultCrypto` struct. The fallback values state which algorithms and keys should be loaded at startup to support old data.

Reply:
Ah, okay.  Works for me.